src/rows-data.cpp

```diff
@@ -0,0 +1,121 @@
+#include <Rcpp.h>
+#include <dplyr.h>
+#include "utils.h"
+#include "rows-data.h"
+
+namespace Slices {
+
+
+CollationType hash_collate(std::string collate) {
+  if (collate == "rows")
+    return rows;
+  else if (collate == "cols")
+    return cols;
+  else
+    return list;
+}
+
+Settings::Settings(Environment execution_env_)
+    : output_colname(as<std::string>(execution_env_[".to"])),
+      include_labels(execution_env_[".labels"]) {
+  collation = hash_collate(as<std::string>(execution_env_[".collate"]));
+}
+
+
+Labels::Labels(Environment execution_env_)
+    : are_unique(execution_env_[".unique_labels"]),
+      slicing_cols(execution_env_[".slicing_cols"]),
+      labels_(execution_env_[".labels_cols"]),
+      n_labels_(Rf_length(execution_env_[".labels_cols"])) {
+}
+
+void Labels::remove(std::vector<int>& index) {
```

Comment:
I like the idea of making the syntax reveal mutation at the call site, but this does not seem very conventional. See for example https://isocpp.org/wiki/faq/references#refs-vs-ptrs

> Old line C programmers sometimes don’t like references since they provide reference semantics that isn’t explicit in the caller’s code. After some C++ experience, however, one quickly realizes this is a form of information hiding, which is an asset rather than a liability. E.g., programmers should write code in the language of the problem rather than the language of the machine.

Plus I feel like I would be guilty not checking for NULL pointers. And it does make for a heavier syntax with all the dereferencing.

How about using references but with this form:
```C++
obj& modify(obj&) { return obj };

void some_function() {
  obj = modify(obj);
}
```

Then it does not matter whether we have value or reference semantics, it becomes an implementation detail.